server/datastore/mysql/software.go

```diff
@@ -520,6 +548,59 @@ func (ds *Datastore) CountSoftware(ctx context.Context, opt fleet.SoftwareListOp
 	return countSoftwareDB(ctx, ds.reader, nil, opt)
 }
 
+// ListVulnerableSoftwareBySource lists all the vulnerable software that matches the given source.
+func (ds *Datastore) ListVulnerableSoftwareBySource(ctx context.Context, source string) ([]fleet.SoftwareWithCPE, error) {
+	var softwareCVEs []struct {
+		fleet.Software
+		CPE  uint   `db:"cpe_id"`
+		CVEs string `db:"cves"`
+	}
+	if err := sqlx.SelectContext(ctx, ds.reader, &softwareCVEs, `
+		SELECT s.*, scv.cpe_id, GROUP_CONCAT(scv.cve SEPARATOR ',') as cves
+		FROM software s
+		JOIN software_cpe scp ON (s.id=scp.software_id)
+		JOIN software_cve scv ON (scp.id=scv.cpe_id)
+		WHERE s.source = ?
+		GROUP BY scv.cpe_id
+	`, source); err != nil {
+		return nil, ctxerr.Wrapf(ctx, err, "listing vulnerable software by source")
+	}
+	software := make([]fleet.SoftwareWithCPE, 0, len(softwareCVEs))
+	for _, sc := range softwareCVEs {
+		for _, cve := range strings.Split(sc.CVEs, ",") {
+			sc.Software.Vulnerabilities = append(sc.Software.Vulnerabilities, fleet.SoftwareCVE{
+				CVE:         cve,
+				DetailsLink: fmt.Sprintf("https://nvd.nist.gov/vuln/detail/%s", cve),
+			})
+		}
+		software = append(software, fleet.SoftwareWithCPE{
+			Software: sc.Software,
+			CPE:      sc.CPE,
+		})
+	}
+	return software, nil
+}
+
+// DeleteVulnerabilitiesByCPECVE deletes the given list of vulnerabilities identified by CPE+CVE.
+func (ds *Datastore) DeleteVulnerabilitiesByCPECVE(ctx context.Context, vulnerabilities []fleet.SoftwareVulnerability) error {
+	if len(vulnerabilities) == 0 {
+		return nil
+	}
+
+	sql := fmt.Sprintf(
+		`DELETE FROM software_cve WHERE (cpe_id, cve) IN (%s)`,
+		strings.TrimSuffix(strings.Repeat("(?,?),", len(vulnerabilities)), ","),
```

Comment:
nitpick, this is a little confusing. Consider using `strings.Join` instead of stripping the extra `,`

```suggestion
		strings.Join(strings.Repeat("(?,?)", len(vulnerabilities)), ","),
```